autoscaling.go

```diff
@@ -0,0 +1,174 @@
+	Time        time.Time `json:"Time"`
+	GroupName   string    `json:"AutoScalingGroupName"`
+	InstanceID  string    `json:"EC2InstanceId"`
+	ActionToken string    `json:"LifecycleActionToken"`
+	Transition  string    `json:"LifecycleTransition"`
+	HookName    string    `json:"LifecycleHookName"`
+}
+
+// NewAutoscalingListener ...
+func NewAutoscalingListener(instanceID string, queue *Queue, autoscaling *autoscaling.AutoScaling) *AutoscalingListener {
+	return &AutoscalingListener{
+		listenerType: "autoscaling",
+		instanceID:   instanceID,
+		queue:        queue,
+		autoscaling:  autoscaling,
+	}
+}
+
+// AutoscalingListener ...
+type AutoscalingListener struct {
+	listenerType string
+	instanceID   string
+	queue        *Queue
+	autoscaling  *autoscaling.AutoScaling
+}
+
+// Type returns a string describing the listener type.
+func (l *AutoscalingListener) Type() string {
+	return l.listenerType
+}
+
+// Start the autoscaling lifecycle hook listener.
+func (l *AutoscalingListener) Start(ctx context.Context, notices chan<- Notice) error {
+	if err := l.queue.Create(); err != nil {
+		return err
+	}
+	defer func() {
+		if err := l.queue.Delete(); err != nil {
+			log.WithError(err).Error("Failed to delete queue")
+		}
+	}()
+
+	if err := l.queue.Subscribe(); err != nil {
+		return err
+	}
+	defer func() {
+		if err := l.queue.Unsubscribe(); err != nil {
+			log.WithError(err).Error("Failed to unsubscribe from sns topic")
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			messages, err := l.queue.GetMessages(ctx)
+			if err != nil {
+				log.WithError(err).Warn("Failed to get messages from SQS")
+			}
+			for _, m := range messages {
+				var env Envelope
+				var msg Message
+
+				if err := l.queue.DeleteMessage(ctx, aws.StringValue(m.ReceiptHandle)); err != nil {
+					log.WithError(err).Warn("Failed to delete message")
+				}
+
+				// unmarshal outer layer
+				if err := json.Unmarshal([]byte(*m.Body), &env); err != nil {
+					log.WithError(err).Warn("Failed to unmarshal envelope")
+					continue
+				}
+
+				log.WithFields(log.Fields{
+					"type":    env.Type,
+					"subject": env.Subject,
+				}).Debug("Received an SQS message")
+
+				// unmarshal inner layer
+				if err := json.Unmarshal([]byte(env.Message), &msg); err != nil {
+					log.WithError(err).Warn("Failed to unmarshal autoscaling message")
+					continue
+				}
+
+				if msg.InstanceID != l.instanceID {
+					log.WithField("target", msg.InstanceID).Debug("Skipping autoscaling event, doesn't match instance id")
+					continue
+				}
+
+				if msg.Transition != "autoscaling:EC2_INSTANCE_TERMINATING" {
+					log.WithField("transition", msg.Transition).Debug("Skipping autoscaling event, not a termination notice")
+					continue
+				}
+
+				notices <- &autoscalingNotice{
+					noticeType:  l.Type(),
+					message:     &msg,
+					autoscaling: l.autoscaling,
+				}
+				return nil
+			}
+		}
+	}
+}
+
+type autoscalingNotice struct {
```

Comment:
Maybe `lifecycleTerminateNotice`?

Reply:
Hm, would `autoscalingTerminationNotice` work? If we go `lifecycleTerminationNotice` I think we should rename the listener to `LifecycleListener`, but that feels less "clear" to me?